aten/src/ATen/native/cuda/MultinomialKernel.cu

```diff
@@ -415,26 +409,27 @@ void multinomial_kernel_impl(Tensor& result, const Tensor& self, const int64_t n
       std::pair<uint64_t, uint64_t> rng_engine_inputs;
 
       if (with_replacement) {
+        // Binary search is warp divergent (so effectively we're running
+        // with just a single thread), but for better utilization,
+        // we need each block to have at least 4 warps.
+        dim3 block(128);
+
+        // Each block will generate a sample from one
+        // distribution concurrently.
+        int grid_y=std::min<int>(numDist, at::cuda::getCurrentDeviceProperties()->maxGridSize[1]);
+        dim3 grid((n_sample-1)/block.x+1, grid_y);
         {
           // See Note [Acquire lock when using random generators]
           std::lock_guard<std::mutex> lock(gen->mutex_);
 
-          // each thread will utilize one random, however, since we have to use
+          // each thread will utilize several randoms (numdist/numblocks*numsamples/numthreads), however, since we have to use
           // curand_uniform4 (See Note [Register spilling in curand call for CUDA < 10]),
-          // offset is 4.
-          rng_engine_inputs = gen->philox_engine_inputs(4);
+          // offset is 4 times that.
+          auto offset = (((n_sample-1)/block.x+1)*((numDist-1)/grid.x+1))*4;
```

Comment:
So what, this is `grid.x * ((numDist-1)/grid.x+1)*4`... so is this just a really longwinded way of saying `numDist * 4`? Or maybe with some extra slop at the end?

Reply:
Oh you are right, I messed this up, changed the kernel but did not change this. Fill fix now.

Reply:
I fixed it and added a comment that hopefully makes things clearer (it's also simpler than it used to be, because each thread is using just 1 random in most cases)